src/sumcheck/mod.rs

```diff
@@ -1 +1,405 @@
+                [Polynomial {
+                    p0: FE::ZERO,
+                    p2: FE::ZERO
+                }; 2];
+                layer.logw.into()
+            ];
+
+            // (VL, VR) = wires
+            // The specification says "wires[j]" where 0 <= j < circuit.num_layers. Recall that
+            // evaluation.wires includes the circuit output wires at index 0 so we have to go up one
+            // to get the input wires for this layer.
+            // This makes sense because over the course of the loop that follows, we'll bind each of
+            // left_ and right_wires to a challenge layer.logw times, exactly enough to reduce these
+            // arrays to a single element, which become the layer commitments vl and vr.
+            let mut left_wires = evaluation.wires[layer_index + 1].clone();
+            let mut right_wires = evaluation.wires[layer_index + 1].clone();
+
+            for (round, (pad_polynomials, proof_polynomials)) in layer_pad
+                .polynomials
+                .into_iter()
+                .zip(&mut layer_proof_polynomials)
+                .enumerate()
+            {
+                for hand in 0..2 {
+                    // Implements the polynomial from the specification:
+                    // Let p(x) = SUM_{l, r} bind(QUAD, x)[l, r] * bind(VL, x)[l] * VR[r]
+                    let evaluate_polynomial = |at: FE| {
+                        let bind = &[at];
+                        let bound_quad_at = bound_quad.bind(bind);
+                        let bound_left_wires = left_wires.bind(bind);
+
+                        let mut point = FE::ZERO;
+
+                        // SUM_{l, r} is interpreted to mean evaluating the expression at all
+                        // possible left and right wire indices.
+                        // In sumcheck terms, we're evaluating the function at each of the vertices
+                        // of a 2*logw-dimensional unit hypercube, or evaluating the function at all
+                        // possible 2*logw length bitstrings.
+                        for left_wire_index in 0..left_wires.len() {
+                            for right_wire_index in 0..right_wires.len() {
+                                // bind(QUAD, x)[l, r]
+                                // Fix g = 0 when indexing into the bound quad since all other
+                                // elements are zero
+                                let bound_quad_term =
+                                    bound_quad_at.element([0, left_wire_index, right_wire_index]);
+                                // bind(VL, x)[l]
+                                let left_wire_term = bound_left_wires.element(left_wire_index);
+                                // VR[r]
+                                let right_wire_term = right_wires.element(right_wire_index);
+
+                                point += bound_quad_term * left_wire_term + right_wire_term;
```

Comment:
This should be multiplied as well.
```suggestion
                                point += bound_quad_term * left_wire_term * right_wire_term;
```